torch/_inductor/codegen/triton.py

```diff
@@ -1050,45 +1052,59 @@ def mask_loads(self, mask):
-                size = min(sizes)
-            else:
-                # Should this go here or in TritonPrinter?
-                def print_min(expr):
-                    if len(expr) == 1:
-                        return texpr(expr[0])
-                    else:
-                        return f"min({texpr(expr[0])}, {print_min(expr[1:])})"
-
-                size = print_min(sizes)
-            # The conditions need to be in parens because of Python's operator precedence.
-            # It'd be less # error-prone to use and/or/not, which is suported by triton
-            cond = f"((0 <= {var}) & ({var} < {size}))"
-            cond_print = f"0 <= {var} < {size}"
-            if not isinstance(original_index, sympy.Integer):
-                var_mask = f"({mask})" if "&" in mask else mask
-                var_mask = f" | ~{var_mask}"
+    def indirect_indexing(self, var, size):
+        class IndirectAssertLine(DeferredLineBase):
+            def __init__(self, line, var, mask, size_map):
+                self.var = var
+                self.mask = mask
+                self.line = line
+                self.size_map = size_map
+
+            def __call__(self):
+                # The conditions need to be in parens because of Python's operator precedence.
+                # It'd be less # error-prone to use and/or/not, which is suported by triton
+                size = V.graph.sizevars.simplify(self.size_map[(self.var, self.mask)])
+                cond = f"(0 <= {self.var}) & ({self.var} < {size})"
+                cond_print = f"0 <= {self.var} < {size}"
+                if self.mask:
+                    cond = f"({cond}) | ~{self.mask}"
+                return self.line.format(cond=cond, cond_print=cond_print)
+
+            def _new_line(self, line):
+                return IndirectAssertLine(line, self.var, self.mask, self.size_map)
+
+        var_str = str(var)
+
+        if config.triton.assert_indirect_indexing and torch.version.hip is None:
+            mask_vars = list(var.mask_vars)
+            if self._load_mask:
+                mask_vars.append(self._load_mask)
+
+            mask = ""
+            if mask_vars:
+                mask = (
+                    f"{mask_vars[0]}"
+                    if len(mask_vars) == 1
+                    else f"({' & '.join(mask_vars)})"
+                )
+
+            size = self.rename_indexing(size)
+
+            # An assertion line may have been written already, if so just
+            # update the max size.
+            size_map = self.indirect_max_sizes
+            map_key = (var_str, mask)
+            existing_size = size_map.get(map_key)
+            if existing_size is None:
+                line = 'tl.device_assert({cond}, "index out of bounds: {cond_print}")'
+                size_map[map_key] = size
+                self.compute.writeline(
+                    IndirectAssertLine(line, var_str, mask, size_map)
```

Comment:
Okay, here is a completely different strategy for de-duplicating the assert lines.  I delay the final formatting of `cond` and `cond_print` until the entire IR has has been executed, and accumulate size information in `TritonKernel.indirect_max_sizes`. This mapping is keyed by variable name and by mask, so it correctly handles masked loads as well.

This also improves upon the current `main` branch which sometimes generates multiple asserts for the same variable. For example, this python code 
```python
import torch

def f(x, y, ind):
    return x[ind] + y[ind]

x = torch.randn(4, 3, device="cuda")
y = torch.randn(2, 3, device="cuda")
ind = torch.zeros(2, 2, dtype=torch.int64, device="cuda")
cfn = torch.compile(f, dynamic=True)
cfn(x, y, ind)
```

Generates two triton asserts on main,
```python
    tmp0 = tl.load(in_ptr0 + (x1), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 4))) | ~xmask, "index out of bounds: 0 <= tmp0 < 4")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tl.device_assert((((0 <= tmp0) & (tmp0 < 2))) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```


But with this version it generates only one:
```python
    tl.device_assert(((0 <= tmp0) & (tmp0 < 2)) | ~xmask, "index out of bounds: 0 <= tmp0 < 2")
    tmp1 = tl.load(in_ptr1 + (x0 + (3*tmp0)), xmask)
    tmp2 = tl.load(in_ptr2 + (x0 + (3*tmp0)), xmask)
```
So it doesn't work when we generate the I don't really know how much value this is adding.